include/RAJA/pattern/forall.hpp

```diff
@@ -427,34 +514,42 @@ RAJA_INLINE void forall(Args&&... args)
  * this reduces implementation overhead and perfectly forwards all arguments
  */
 template <typename ExecutionPolicy, typename... Args>
-RAJA_INLINE void forall_Icount(Args&&... args)
+RAJA_INLINE concepts::enable_if_t<resources::EventProxy, type_traits::is_execution_policy<ExecutionPolicy>> forall_Icount(Args&&... args)
+{
+  auto r = resources::get_default_resource(ExecutionPolicy());
+  return forall_Icount<ExecutionPolicy>(r, std::forward<Args>(args)...);
+}
+template <typename ExecutionPolicy, typename... Args>
+RAJA_INLINE resources::EventProxy forall_Icount(resources::Resource &r, Args&&... args)
 {
-  policy_by_value_interface::forall_Icount(ExecutionPolicy(), std::forward<Args>(args)...);
+  return policy_by_value_interface::forall_Icount(r, ExecutionPolicy(), std::forward<Args>(args)...);
 }
 
 namespace detail
 {
 
 template <typename T, typename ExecutionPolicy, typename LoopBody>
-RAJA_INLINE void CallForall::operator()(T const& segment,
-                                        ExecutionPolicy,
-                                        LoopBody body) const
+RAJA_INLINE camp::resources::EventProxy CallForall::operator()(T const& segment,
+                                                               ExecutionPolicy,
+                                                               LoopBody body,
+                                                               camp::resources::Resource &r) const
 {
   // this is only called inside a region, use impl
   using policy::sequential::forall_impl;
   RAJA_FORCEINLINE_RECURSIVE
-  forall_impl(ExecutionPolicy(), segment, body);
+  return forall_impl(r, ExecutionPolicy(), segment, body);
```

Comment:
I've been imagining this with the policy before the resource. What do other people's think about one way vs the other?
```
forall<policy>(resource, ...);
forall(policy, resource, ...); vs forall(resource, policy, ...);
```

Reply:
Since it's unlikely that folks will call the 'impl' directly and will instead do forall<policy>(...) specifying the policy as a template parameter and will pass a resource object as an argument, I think it really doesn't matter much. Are you interested in preserving ordering of concepts + args?

Reply:
I should have picked the line with the signature instead of that line. It seems weird to me that the forall with policy by type and the forall with policy by value have take the policy and resource in different orders.